### DISCUSSION TITLE
how to hide the server name for APISIX on the response header

### DISCUSSION THREAD

**Discussion by michaelh1355** [apache/apisix#8011] [General]
hello !

I have already hide the APISIX version on my APISIX server. we need to hide the server name for APISIX on the response header too .please advise how to do this. thanks!

**Comment by tokers** [comment]
See https://github.com/apache/apisix/blob/master/conf/config-default.yaml#L45

**Comment by michaelh1355** [comment]
您好，


enable_server_tokens: true        # Whether the APISIX version number should be shown in Server header.


这个参数能关闭 版本号，但我还想隐藏APISIX标识，请问怎么修改？




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/apisix"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年9月28日(星期三) 下午5:46
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/apisix] how to hide the server name for APISIX on the response header (Discussion #8011)





 
See https://github.com/apache/apisix/blob/master/conf/config-default.yaml#L45

**Reply by tzssangglass** [reply]
Only by modifying the source code, ref: https://github.com/apache/apisix/blob/61afafdd46fe40c03955b6cb058df3e6fad5db45/apisix/init.lua#L641

**Comment by michaelh1355** [comment]
您好，


请问源代码修改APISIX标识，需要修改哪几个参数？谢谢！




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/apisix"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月8日(星期六) 下午2:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/apisix] how to hide the server name for APISIX on the response header (Discussion #8011)





 
Only by modifying the source code, ref: https://github.com/apache/apisix/blob/61afafdd46fe40c03955b6cb058df3e6fad5db45/apisix/init.lua#L641

**Reply by tzssangglass** [reply]
I've pointed you to the source code where to set the APISIX response header, you can just remove that line.

**Reply by tokers** [reply]
Why not just use the response-rewrite plugin to rewrite the Server header?

**Comment by michaelh1355** [comment]
您好，


您提供这个解决方案比重装编译apisix更加简单，我只想将响应头的Server: APISIX改成Server: APP或者删除&nbsp;Server: APISIX标识，如何编写response-rewrite plugin&nbsp;配置文件？ 我看了帮助文档（https://apisix.apache.org/zh/docs/apisix/plugins/response-rewrite），始终还是没看明白语法怎么编写。


感谢！






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/apisix"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月8日(星期六) 晚上7:56
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/apisix] how to hide the server name for APISIX on the response header (Discussion #8011)





 
Why not just use the response-rewrite plugin to rewrite the Server header?

**Reply by tokers** [reply]
Then you need to learn from the getting-started and the core concepts, before you actually configure this plugin.

**Comment by michaelh1355** [comment]
APISIX标识可以隐藏了，感谢您的建议。









------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/apisix"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月9日(星期天) 上午9:52
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/apisix] how to hide the server name for APISIX on the response header (Discussion #8011)





 
Then you need to learn from the getting-started and the core concepts, before you actually configure this plugin.

**Comment by SilentEntity** [comment]
one can edit the value for `Server` header in response by using global_rules.

Below is the example for standalone mode yaml configuration:
```
global_rules:
      - id: 1
        plugins:
          response-rewrite:
            headers:
              set:
                Server: "we purple you"
```

If someone using the admin:
```
curl http://127.0.0.1:9180/apisix/admin/routes/1 \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "methods": ["GET"],
    "uri": "/test/index.html",
    "plugins": {
        "response-rewrite": {
            "body": "{\"code\":\"ok\",\"message\":\"new json body\"}",
            "headers": {
                "set": {
                    "Server": "we purple you"
                }
            },
            "vars":[
                [ "status","==",200 ]
            ]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:80": 1
        }
    }
}'
```

**Reply by grlicky** [reply]
Just for a reference, the `Server` header can be removed completely using the `headers.remove` attribute of the response-rewrite plugin:

```
global_rules:
      - id: 1
        plugins:
          response-rewrite:
            headers:
              remove:
                - Server
```

**Comment by cbbntla** [comment]
apisix set response header `Server`  at `header_filter` phase.
> https://github.com/apache/apisix/blob/61afafdd46fe40c03955b6cb058df3e6fad5db45/apisix/init.lua#L641

I tried to re-set `Server` header with `serverless-post-function` plugin, choose `header_filter` phase, and it works.
```
{
  "plugins": {
    "serverless-post-function": {
      "functions": [
        "return function() local core = require(\"apisix.core\"); core.response.set_header(\"Server\", \"foo bar\") end"
      ],
      "phase": "header_filter"
    }
  }
}
```

![image](https://user-images.githubusercontent.com/66546940/233305388-b31cfc2f-e423-46b1-9ed1-93ba2b8837b0.png)
